File: lab-003/python/exposure_fusion/crop.py
"""Largest hole-free common crop for aligned exposures."""

from __future__ import annotations

from collections.abc import Sequence

import numpy as np

from .contracts import CropRect
from .errors import FusionError


def _largest_rectangle(mask: np.ndarray) -> CropRect:
    heights = np.zeros(mask.shape[1], dtype=np.int32)
    best = CropRect(0, 0, 0, 0)
    for row in range(mask.shape[0]):
        heights = np.where(mask[row], heights + 1, 0)
        stack: list[tuple[int, int]] = []
        for column in range(mask.shape[1] + 1):
            height = int(heights[column]) if column < mask.shape[1] else 0
            start = column
            while stack and stack[-1][1] > height:
                index, previous_height = stack.pop()
                area = previous_height * (column - index)
                if area > best.width * best.height:
                    best = CropRect(index, row - previous_height + 1, column - index, previous_height)
                start = index
            if not stack or stack[-1][1] < height:
                stack.append((start, height))
    return best


def crop_common_region(
    masks: Sequence[np.ndarray],
    *,
    inset: int = 2,
    maximum_loss: float = 0.20,
) -> CropRect:
    if len(masks) != 3:
        raise FusionError("INVALID_IMAGE_COUNT", "Three validity masks are required.")
    common = np.logical_and.reduce([mask > 0 for mask in masks])
    rect = _largest_rectangle(common)
    rect = CropRect(
        rect.x + inset,
        rect.y + inset,
        max(0, rect.width - inset * 2),
        max(0, rect.height - inset * 2),
    )
    total = common.shape[0] * common.shape[1]
    if rect.width <= 0 or rect.height <= 0 or rect.width * rect.height < total * (1 - maximum_loss):
        raise FusionError("EXCESSIVE_CROP", "Alignment would discard more than 20% of the frame.")
    return rect

File: lab-003/web/js/crop.js
import { FusionError } from "./errors.js";

export function largestRectangle(mask, width, height) {
  const heights = new Int32Array(width);
  let best = { x: 0, y: 0, width: 0, height: 0 };
  for (let row = 0; row < height; row += 1) {
    for (let column = 0; column < width; column += 1) {
      heights[column] = mask[row * width + column] ? heights[column] + 1 : 0;
    }
    const stack = [];
    for (let column = 0; column <= width; column += 1) {
      const current = column < width ? heights[column] : 0;
      let start = column;
      while (stack.length && stack.at(-1).height > current) {
        const item = stack.pop();
        const candidate = {
          x: item.start,
          y: row - item.height + 1,
          width: column - item.start,
          height: item.height,
        };
        if (candidate.width * candidate.height > best.width * best.height) best = candidate;
        start = item.start;
      }
      if (!stack.length || stack.at(-1).height < current) stack.push({ start, height: current });
    }
  }
  return best;
}

export function cropCommonRegion(masks, width, height, inset = 2) {
  if (masks.length !== 3) throw new FusionError("INVALID_IMAGE_COUNT", "Three masks are required.");
  const common = new Uint8Array(width * height);
  for (let pixel = 0; pixel < common.length; pixel += 1) {
    common[pixel] = masks.every((mask) => mask[pixel]) ? 255 : 0;
  }
  const largest = largestRectangle(common, width, height);
  const crop = {
    x: largest.x + inset,
    y: largest.y + inset,
    width: Math.max(0, largest.width - inset * 2),
    height: Math.max(0, largest.height - inset * 2),
  };
  if (!crop.width || !crop.height || crop.width * crop.height < width * height * 0.8) {
    throw new FusionError("EXCESSIVE_CROP", "Alignment would discard more than 20% of the frame.");
  }
  return crop;
}

export function cropImage(image, crop) {
  const data = new Uint8ClampedArray(crop.width * crop.height * 4);
  for (let row = 0; row < crop.height; row += 1) {
    const sourceStart = ((crop.y + row) * image.width + crop.x) * 4;
    data.set(image.data.subarray(sourceStart, sourceStart + crop.width * 4), row * crop.width * 4);
  }
  return { width: crop.width, height: crop.height, data };
}
